Pcap++/header/PcapFilter.h

```diff
@@ -484,82 +484,117 @@ namespace pcpp
+		void setFilters(const std::vector<GeneralFilter*>& filters);
 
-		void parseToString(std::string& result);
+		/**
+		 * Remove all filters from the composite filter.
+		 */
+		void clearAllFilters() { m_FilterList.clear(); }
 	};
 
-
-
 	/**
-	 * @class OrFilter
-	 * A class for connecting several filters into one filter with logical "or" between them. For example: if the 2 filters are: "IPv4 address =
-	 * x.x.x.x" + "TCP port dst = 80", then the new filter will be: "IPv4 address = x.x.x.x _OR_ TCP port dst = 80"<BR>
-	 * This class follows the composite design pattern<BR>
-	 * For deeper understanding of the filter concept please refer to PcapFilter.h
-	 * @todo add some methods: "addFilter", "removeFilter", "clearAllFilter"
+	 * Supported composite logic filter operators enum
 	 */
-	class OrFilter : public GeneralFilter
+	enum class CompositeLogicFilterOp
 	{
-	private:
-		std::vector<GeneralFilter*> m_FilterList;
-	public:
-
-		/**
-		 * An empty constructor for this class. Use addFilter() to add filters to the or condition
-		 */
-		OrFilter() {}
+		/** Logical AND operation */
+		AND,
+		/** Logical OR operation */
+		OR,
+	};
 
+	namespace detail
+	{
+		/* Could potentially be moved into CompositeLogicFilter as a private member function, with if constexpr when C++17 is the minimum supported standard.*/
 		/**
-		 * A constructor that gets a list of pointers to filters and creates one filter from all filters with logical "or" between them
-		 * @param[in] filters The list of pointers to filters
+		 * Returns the delimiter for joining filter strings for the composite logic filter operation.
+		 * @return A string li
 		 */
-		explicit OrFilter(std::vector<GeneralFilter*>& filters);
+		template <CompositeLogicFilterOp op> constexpr const char *getCompositeLogicOpDelimiter() = delete;
+		template <> constexpr const char *getCompositeLogicOpDelimiter<CompositeLogicFilterOp::AND>() { return " and "; };
+		template <> constexpr const char *getCompositeLogicOpDelimiter<CompositeLogicFilterOp::OR>() { return " or "; };
+	}
 
-		/**
-		 * Add filter to the or condition
-		 * @param[in] filter The filter to add
-		 */
-		void addFilter(GeneralFilter* filter) { m_FilterList.push_back(filter); }
+	/**
+	 * @class CompositeLogicFilter
+	 * A class for connecting several filters into one filter with logical operation between them.<BR>
+	 * For deeper understanding of the filter concept please refer to PcapFilter.h
+	 */
+	template <CompositeLogicFilterOp op>
+	class CompositeLogicFilter : public CompositeFilter
+	{
+	public:
+		using CompositeFilter::CompositeFilter;
 
-		void parseToString(std::string& result);
+		void parseToString(std::string& result)
+		{
+			result.clear();
+			for (auto it = m_FilterList.cbegin(); it != m_FilterList.cend(); ++it)
+			{
+				std::string innerFilter;
+				(*it)->parseToString(innerFilter);
+				result += '(' + innerFilter + ')';
+				if (m_FilterList.back() != *it)
```

Comment:
This has the edge case of technically being false  in the middle of a run if a filter is added twice to the composite filter as it compares the pointer values and not the iterators. Though unsure if the change has a performance impact.

```suggestion
				if (m_FilterList.cend() - 1 != it)
```

Reply:
could you give an example demonstrating the issue you are concerned about?

Reply:
Essentially that this test will fail with the current implementation.

```
pcpp::IPFilter ipFilter("10.0.0.6", pcpp::SRC);
protoFilter.setProto(pcpp::UDP);
std::vector<pcpp::GeneralFilter*> filterVec;
filterVec.push_back(&ipFilter);
filterVec.push_back(&protoFilter);

pcpp::CompositeLogicFilter<pcpp::CompositeLogicFilterOp::AND> compositeFilter(filterVec);

// Adding the same proto filter again to the composite filter.
compositeFilter.addFilter(&protoFilter);
compositeFilter.parseToString(filterAsString);
PTF_ASSERT_EQUAL(filterAsString, "(ip and src net 10.0.0.6) and (udp) and (udp)")
```

TestPcapFiltersOffline             : FAILED (\<ProjectLocation\>\PcapPlusPlus\Tests\Pcap++Test\Tests\FilterTests.cpp:690). Assert EQUAL failed:
   Actual:   (ip and src net 10.0.0.6) and (udp)(udp)
   Expected: (ip and src net 10.0.0.6) and (udp) and (udp)

Reply:
I am thinking maybe the correct way to handle this is to check duplicate or not?

Reply:
Not sure, if checking for duplicates is the better option, as it would end up being a larger change for dubious benefit over the suggested change. As I said, this is an edge case. I assume most sane programs wouldn't wouldn't be adding the same filter twice regularly.

When checking for duplicates, you either have to change the backing structure from `vector` to `set`, which has an overall slower iteration speed than `vector`. Or you check for duplicates on entry keeping the vector, making `addFilter` an O(n) worst case instead of O(1) amortized.

In my opinion it is better to eat the potential performance penalty of parsing the same filter twice then adding overhead on all iterations/additions of filters by checking if there is a duplicate, since most of the time there won't be any duplicates in sane filters in the first place.

Reply:
sure, I think it's fine, I think your current solution makes sense.
could you try to add more edge cases to verify the correctness of the behavior?

Reply:
I think the current implementation is good enough 👍

Reply:
Fix implemented ebd8466